### ISSUE TITLE
Map multibindings does not support Lazy

### ISSUE THREAD

**Issue by cckroets** [google/dagger#1630]
Currently Map multibindings can provide:
Map<Key, Value>
Map<Key, Provider<Value>>

I would like it to also provide:
Map<Key, Lazy<Value>>

Right now we use the Provider method. However our map only needs one instance of each Value. If we scope the Value we get one instance. If we leave Values un-scoped we get a new instance every time we access the map. Ideally each Value could be left un-scoped since they contain no state, but now we are creating many new instances of these objects unnecessarily.

**Comment by Chang-Eric** [comment]
We'll need to think about the semantics of this. Right now for instance I think we'll always use the same map for `Map<K, V>` and `Map<K, Provider<V>>`. This would likely need to be different to give you a different map per injection, which may be confusing or complicate the code. Not that it is impossible to do but more just that it adds complexity for something that is likely not used that much. We'll have to consider the tradeoffs to see if this feature is worth it.

**Comment by bcorso** [comment]
Ah, sorry! I closed this since the bug was pretty old and not upvoted -- I didn't spend much time on it. (Also that unit test is way to long :-P)

Anyway, I took a look and I agree there is a bug here.

The most concise example I can think of is this:

```java
@Component(modules = MyModule.class)
public interface MyComponent {
    Map<String, Lazy<? extends X>> getXMap();
}

@Module
interface MyModule {
    @Binds @IntoMap @StringKey("x1")
    Lazy<? extends X> x1(Lazy<X1> impl);

    @Provides @IntoMap @StringKey("x2")
    static Lazy<? extends X> x2(Lazy<X2> impl) {
        return impl;
    }
}
```

You would expect the `@Binds` and `@Provides` above to behave the same, however the `@Provides` returns a lazy factory but the `@Binds` just casts to `Lazy` (which is clearly wrong):

```java
public final class DaggerMyComponent implements MyComponent {
  @Override
  public Map<String, Lazy<? extends X>> xMap() {
    return MapBuilder.<String, Lazy<? extends MyComponent.X>>newMapBuilder(2)
        .put("x1", (Lazy<? extends X>) new X1())
        .put("x2", MyModule_X2Factory.x2(DoubleCheck.lazy(X2_Factory.create())))
        .build();
  }
}
```

There's a couple ways you could fix this. From your post, I think you'd prefer that @Binds behaves like the @Provides above, however IMO that would also be a bug.

In particular, returning framework types (e.g. Provider or Lazy) is not allowed for normal bindings:

```java
@Module
interface MyModule {
    // error: @Binds methods must not return framework types
    @Binds Lazy<? extends X> x1(Lazy<X1> impl);

    // error: @Provides methods must not return framework types
    @Provides Lazy<? extends X> x2(Lazy<X2> impl) {
      return impl;
    }
}
```

So I think it's a bug that it was ever allowed for multibinding contributions., and we should make it fail for consitency.

However, I'm assuming you're probably relying on the current behavior of the `@Provides Lazy` multibinding contribution so that would probably leave you in a bad place. 

Instead, what I think what you really want is for us to implement https://github.com/google/dagger/issues/1630 so that you can contribute type `X` and inject `Map<String, Lazy<X>>`.

Please let me know if that all sounds right to you.

**Comment by bcorso** [comment]
>  If we scope the Value we get one instance. Ideally each Value could be left un-scoped since they contain no state, but now we are creating many new instances of these objects unnecessarily.

Just to throw it out there in case it helps someone, this is essentially what `@Reusable` was created for. It's a bit more work to scope each multibinding contribution, but it should be cheaper than Lazy since it doesn't require double-check-locking and it can cache across multiple classes.

**Comment by hadilq** [comment]
Thanks @bcorso for explanation! IMO, not providing `Map<Key, Lazy<Value>>` is just an inconsistency in the dagger API. My use case in https://github.com/google/dagger/issues/2474 was gathering some mapper classes in one parent mapper class to convert some types. I used their types as the key in the `Map`. In my case, it's efficient to initiate the mapper classes once per their parent, and keep it for future usage, also avoiding initiation of mappers that are not needed, which is the definition of `Lazy`. Clearly I cannot do it with `@Reusable`.

**Comment by bcorso** [comment]
> IMO, not providing Map<Key, Lazy<Value>> is just an inconsistency in the dagger API. My use case in #2474 was gathering some mapper classes in one parent mapper class to convert some types.

Yeah, I think we mostly agree on that. However, this particular feature is not high priority at the moment so just wanted to mention some possible work arounds.

> In my case, it's efficient to initiate the mapper classes once per their parent, and keep it for future usage, also avoiding initiation of mappers that are not needed, which is the definition of Lazy. Clearly I cannot do it with @Reusable.

@hadilq, just curious, is there a particular constraint that prevents you from using `Map<Key, Provider<Value>>` with `@Reusable` on each `Value`? Sounds like it would get you similar behavior to what you're asking for.

**Comment by hadilq** [comment]
> just curious, is there a particular constraint that prevents you from using Map<Key, Provider<Value>> with @Reusable (or actual scope) on each Value? Sounds like it would get you similar behavior to what you're asking for.

@bcorso We are trying to reduce usage of `@Reusabe` in our project, because they act similar to singletons. Also the module, I was working on, has a dagger component that didn't provide scopes, so I tried to not touching everything to provide an efficient solution you suggested. But in the end, you are right, and we need to refactor our code to achieve that.
